[SPARK-13012] [Documentation] Replace example code in ml-guide.md using include_example

devaraj-kavali opened the pull request:
Replaced example code in ml-guide.md using include_example

Commit:
[SPARK-13012] [Documentation] Replace example code in ml-guide.md using
include_example

Replaced example code in ml-guide.md using include_example

AmplabJenkins commented:
Can one of the admins verify this patch?

yinxusen commented:
Thanks @devaraj-kavali for taking this. I'll check it later today.

yinxusen commented:
ok to test

yinxusen commented:
test it please

yinxusen commented:
@mengxr Can you help me calling Jenkins to test it?

mengxr commented:
ok to test

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50765/
Test FAILed.

yinxusen commented:
retest this please

SparkQA commented:
**[Test build #50788 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50788/consoleFull)** for PR 11053 at commit [`d485a7d`](https://github.com/apache/spark/commit/d485a7dca97d25e8cd0dc380029112fd17c569ff).

SparkQA commented:
**[Test build #50788 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50788/consoleFull)** for PR 11053 at commit [`d485a7d`](https://github.com/apache/spark/commit/d485a7dca97d25e8cd0dc380029112fd17c569ff).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class JavaEstimatorTransformerParamExample `
  * `class Document1 implements Serializable `
  * `class LabeledDocument1 extends Document1 implements Serializable `
  * `public class JavaModelSelectionViaCrossValidationExample `
  * `public class JavaModelSelectionViaTrainValidationSplitExample `
  * `class Document implements Serializable `
  * `class LabeledDocument extends Document implements Serializable `
  * `public class JavaPipelineExample `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50788/
Test FAILed.

yinxusen commented:
@devaraj-kavali, pls fix the style error first.

Commit:
Fixed the style errors and corrected .py files

devaraj-kavali commented:
I have fixed the errors, @yinxusen Please have a look into this.

SparkQA commented:
**[Test build #50819 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50819/consoleFull)** for PR 11053 at commit [`1373996`](https://github.com/apache/spark/commit/13739964ee9f0cdaa263c2067d68f4a6ee736f3a).

SparkQA commented:
**[Test build #50819 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50819/consoleFull)** for PR 11053 at commit [`1373996`](https://github.com/apache/spark/commit/13739964ee9f0cdaa263c2067d68f4a6ee736f3a).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50819/
Test FAILed.

Commit:
Fixed python style check warnings

SparkQA commented:
**[Test build #50824 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50824/consoleFull)** for PR 11053 at commit [`6cc98c9`](https://github.com/apache/spark/commit/6cc98c9fc798082d5ac22800ffdda7593290019d).

SparkQA commented:
**[Test build #50824 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50824/consoleFull)** for PR 11053 at commit [`6cc98c9`](https://github.com/apache/spark/commit/6cc98c9fc798082d5ac22800ffdda7593290019d).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50824/
Test FAILed.

Commit:
Fixed scala warning in this example

SparkQA commented:
**[Test build #50825 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50825/consoleFull)** for PR 11053 at commit [`48cafb1`](https://github.com/apache/spark/commit/48cafb1f2678d7a015c56735c61e94c70584f2eb).

SparkQA commented:
**[Test build #50825 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50825/consoleFull)** for PR 11053 at commit [`48cafb1`](https://github.com/apache/spark/commit/48cafb1f2678d7a015c56735c61e94c70584f2eb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50825/
Test PASSed.

yinxusen commented on examples/src/main/java/org/apache/spark/examples/mllib/JavaEstimatorTransformerParamExample.java:
```diff
@@ -0,0 +1,108 @@
+ * limitations under the License.
+ */
+
+import java.util.Arrays;
```
Use example on and off to imports, too. E.g.

```scala
// $example on$
import org.apache.spark.ml.classification.LogisticRegressionModel;
import org.apache.spark.ml.param.ParamMap;
...
// $example off$

yinxusen commented on examples/src/main/java/org/apache/spark/examples/mllib/JavaEstimatorTransformerParamExample.java:
```diff
@@ -0,0 +1,108 @@
+ */
+
+import java.util.Arrays;
+import java.util.List;
```
delete this import since Java List is not used in the example.

yinxusen commented on examples/src/main/java/org/apache/spark/examples/mllib/JavaEstimatorTransformerParamExample.java:
```diff
@@ -0,0 +1,108 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
Add package here: 

```scala
package org.apache.spark.examples.ml;
```

yinxusen commented:
@devaraj-kavali Please keep 2-indent for Java code.

yinxusen commented on examples/src/main/java/org/apache/spark/examples/mllib/JavaPipelineExample.java:
```diff
@@ -0,0 +1,123 @@
+
+// Labeled and unlabeled instance types.
+// Spark SQL can infer schema from Java Beans.
+class Document implements Serializable {
```
Let's define `Document` in a single place. Try to merge it with class `Document1`.

Commit:
Fixed the review comments

SparkQA commented:
**[Test build #50914 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50914/consoleFull)** for PR 11053 at commit [`3a871c1`](https://github.com/apache/spark/commit/3a871c1afd9d98cd5fee970266bfcd4e5e407947).

SparkQA commented:
**[Test build #50914 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50914/consoleFull)** for PR 11053 at commit [`3a871c1`](https://github.com/apache/spark/commit/3a871c1afd9d98cd5fee970266bfcd4e5e407947).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class Document implements Serializable `
  * `public class LabeledDocument extends Document implements Serializable `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50914/
Test PASSed.

devaraj-kavali commented:
Thanks @yinxusen for review, I have fixed the comments, Please have a look into this.

yinxusen commented:
@devaraj-kavali All Java code should be 2-indent, other than 4-indent.

Commit:
Changed the package from mllib to ml and indentation to 2 spaces for all
the Java code

SparkQA commented:
**[Test build #50958 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50958/consoleFull)** for PR 11053 at commit [`50f5fd9`](https://github.com/apache/spark/commit/50f5fd98271b8389290c9cc5124f566850a9d91a).

SparkQA commented:
**[Test build #50958 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50958/consoleFull)** for PR 11053 at commit [`50f5fd9`](https://github.com/apache/spark/commit/50f5fd98271b8389290c9cc5124f566850a9d91a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50958/
Test PASSed.

devaraj-kavali commented:
I am sorry @yinxusen for making you to give the same comment gain, I have fixed the indent issue and moved the files from mllib to ml package.

yinxusen commented:
Never mind, I'll check it later.

2016年2月9日星期二，Devaraj Kavali <notifications@github.com> 写道：

> I am sorry @yinxusen <https://github.com/yinxusen> for making you to give
> the same comment gain, I have fixed the indent issue and moved the files
> from mllib to ml package.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/11053#issuecomment-181752685>.
>


-- 
Cheers
-----------------------------------
Xusen Yin    (尹绪森)
LinkedIn: https://cn.linkedin.com/in/xusenyin

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaEstimatorTransformerParamExample.java:
```diff
@@ -0,0 +1,118 @@
+import org.apache.spark.sql.DataFrame;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SQLContext;
+
```
remove the blank line

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaEstimatorTransformerParamExample.java:
```diff
@@ -0,0 +1,118 @@
+    // $example on$
+    // Prepare training data.
+    // We use LabeledPoint, which is a JavaBean. Spark SQL can convert RDDs of
+    // JavaBeans
```
re-format the comments like:

```scala
// We use LabeledPoint, which is a JavaBean. Spark SQL can convert RDDs of JavaBeans into
// DataFrames, where it uses the bean metadata to infer the schema.
```

Only break lines when reaching 100 characters or meaningful separations.

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaEstimatorTransformerParamExample.java:
```diff
@@ -0,0 +1,118 @@
+    // We use LabeledPoint, which is a JavaBean. Spark SQL can convert RDDs of
+    // JavaBeans
+    // into DataFrames, where it uses the bean metadata to infer the schema.
+    DataFrame training = sqlContext.createDataFrame(Arrays.asList(
```
It will be more clear if we re-format the code like this:

```scala
    DataFrame training = sqlContext.createDataFrame(
      Arrays.asList(
        new LabeledPoint(1.0, Vectors.dense(0.0, 1.1, 0.1)),
        new LabeledPoint(0.0, Vectors.dense(2.0, 1.0, -1.0)),
        new LabeledPoint(0.0, Vectors.dense(2.0, 1.3, 1.0)),
        new LabeledPoint(1.0, Vectors.dense(0.0, 1.2, -0.5))
      ), LabeledPoint.class);
```

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaEstimatorTransformerParamExample.java:
```diff
@@ -0,0 +1,118 @@
+    LogisticRegression lr = new LogisticRegression();
+    // Print out the parameters, documentation, and any default values.
+    System.out.println("LogisticRegression parameters:\n" + lr.explainParams()
+      + "\n");
```
Merge the line with the previous one. Only break lines when necessary.

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaEstimatorTransformerParamExample.java:
```diff
@@ -0,0 +1,118 @@
+    // Since model1 is a Model (i.e., a Transformer produced by an Estimator),
+    // we can view the parameters it used during fit().
+    // This prints the parameter (name: value) pairs, where names are unique IDs
+    // for this
```
reformat comments

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaEstimatorTransformerParamExample.java:
```diff
@@ -0,0 +1,118 @@
+    // for this
+    // LogisticRegression instance.
+    System.out.println("Model 1 was fit using parameters: "
+      + model1.parent().extractParamMap());
```
merge into the previous line

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaEstimatorTransformerParamExample.java:
```diff
@@ -0,0 +1,118 @@
+    // We may alternatively specify parameters using a ParamMap.
+    // Specify 1 Param.
+    ParamMap paramMap = new ParamMap().put(lr.maxIter().w(20))
+      .put(lr.maxIter(), 30) // This overwrites the original maxIter.
```
reformat the code. Use 2-indent for inline comments, i.e.

```scala
.put(lr.maxIter(), 30)  // This overwrites the original maxIter.
```

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaEstimatorTransformerParamExample.java:
```diff
@@ -0,0 +1,118 @@
+
+    // One can also combine ParamMaps.
+    ParamMap paramMap2 = new ParamMap().put(lr.probabilityCol().w(
+      "myProbability")); // Change output column name
```
merge with the previous line

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaEstimatorTransformerParamExample.java:
```diff
@@ -0,0 +1,118 @@
+      "myProbability")); // Change output column name
+    ParamMap paramMapCombined = paramMap.$plus$plus(paramMap2);
+
+    // Now learn a new model using the paramMapCombined parameters.
```
reformat comments

yinxusen commented:
@devaraj-kavali Thanks for working on it!

There are some code style issues that I found in the code. Do you use some auto code format tools? It's better to keep the code format with previous code snippets in markdown files. Since it's example code, we want to keep it clear to read for users. Be careful with code style guide, like we use 2-indent for inline comments. When finishing the code transfer, please test each of them to see whether it can run correctly or not.

devaraj-kavali commented:
Thanks for the review @yinxusen. I have configured the code format in IDE and using the same for formatting the code. I will fix these comments and update.

Commit:
Fixed the review comments about the style issues

Commit:
Merge branch 'master' into SPARK-13012

Conflicts:
	docs/ml-guide.md

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51429/
Test FAILed.

SparkQA commented:
**[Test build #51430 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51430/consoleFull)** for PR 11053 at commit [`6b63899`](https://github.com/apache/spark/commit/6b63899622cf918b44bdc6a46afcd118dc3e4add).

SparkQA commented:
**[Test build #51430 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51430/consoleFull)** for PR 11053 at commit [`6b63899`](https://github.com/apache/spark/commit/6b63899622cf918b44bdc6a46afcd118dc3e4add).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51430/
Test PASSed.

srowen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaDocument.java:
```diff
@@ -0,0 +1,52 @@
+public class JavaDocument implements Serializable {
+
+  /** Serializable version identifier */
+  private static final long serialVersionUID = 7090547542135729169L;
```
Generally, it's a bad idea to include `serialVersionUID`. It exists for particular reasons that don't really apply to Spark

srowen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaDocument.java:
```diff
@@ -0,0 +1,52 @@
+    return this.id;
+  }
+
+  public void setId(long id) {
```
I don't think you need setters, but you probably need a no-arg constructor to make this JavaBeans-friendly

srowen commented on examples/src/main/scala/org/apache/spark/examples/ml/ModelSelectionViaCrossValidationExample.scala:
```diff
@@ -0,0 +1,109 @@
+// scalastyle:off println
+package org.apache.spark.examples.ml
+
+import org.apache.spark.{ SparkConf, SparkContext }
```
Trivial, but braces shouldn't have a space inside in all of these imports

Commit:
Removed setters for Java beans and spaces inside imports

SparkQA commented:
**[Test build #51434 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51434/consoleFull)** for PR 11053 at commit [`446d6a4`](https://github.com/apache/spark/commit/446d6a469e865da5a464e906cc1acb9031196603).

SparkQA commented:
**[Test build #51434 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51434/consoleFull)** for PR 11053 at commit [`446d6a4`](https://github.com/apache/spark/commit/446d6a469e865da5a464e906cc1acb9031196603).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51434/
Test PASSed.

devaraj-kavali commented:
Thanks @srowen for review and comments. I have removed serialVersionUID and setters in Java Beans and also addressed the unnecessary spaces between braces in imports.

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaModelSelectionViaCrossValidationExample.java:
```diff
@@ -0,0 +1,120 @@
+      .setMaxIter(10)
+      .setRegParam(0.01);
+    Pipeline pipeline = new Pipeline()
+      .setStages(new PipelineStage[] { tokenizer, hashingTF, lr });
```
Remove spaces in brace like this `{tokenizer, hashingTF, lr}`

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaModelSelectionViaCrossValidationExample.java:
```diff
@@ -0,0 +1,120 @@
+    // With 3 values for hashingTF.numFeatures and 2 values for lr.regParam,
+    // this grid will have 3 x 2 = 6 parameter settings for CrossValidator to choose from.
+    ParamMap[] paramGrid = new ParamGridBuilder()
+      .addGrid(hashingTF.numFeatures(), new int[] { 10, 100, 1000 })
```
ditto

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaModelSelectionViaCrossValidationExample.java:
```diff
@@ -0,0 +1,120 @@
+    // this grid will have 3 x 2 = 6 parameter settings for CrossValidator to choose from.
+    ParamMap[] paramGrid = new ParamGridBuilder()
+      .addGrid(hashingTF.numFeatures(), new int[] { 10, 100, 1000 })
+      .addGrid(lr.regParam(), new double[] { 0.1, 0.01 })
```
ditto

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaModelSelectionViaTrainValidationSplitExample.java:
```diff
@@ -0,0 +1,82 @@
+    DataFrame data = jsql.read().format("libsvm").load("data/mllib/sample_libsvm_data.txt");
+
+    // Prepare training and test data.
+    DataFrame[] splits = data.randomSplit(new double[] { 0.9, 0.1 }, 12345);
```
`{0.9, 0.1}`

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaModelSelectionViaTrainValidationSplitExample.java:
```diff
@@ -0,0 +1,82 @@
+    // TrainValidationSplit will try all combinations of values and determine best model using
+    // the evaluator.
+    ParamMap[] paramGrid = new ParamGridBuilder()
+      .addGrid(lr.regParam(), new double[] { 0.1, 0.01 })
```
ditto

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaModelSelectionViaTrainValidationSplitExample.java:
```diff
@@ -0,0 +1,82 @@
+    ParamMap[] paramGrid = new ParamGridBuilder()
+      .addGrid(lr.regParam(), new double[] { 0.1, 0.01 })
+      .addGrid(lr.fitIntercept())
+      .addGrid(lr.elasticNetParam(), new double[] { 0.0, 0.5, 1.0 })
```
ditto

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaPipelineExample.java:
```diff
@@ -0,0 +1,89 @@
+      .setMaxIter(10)
+      .setRegParam(0.01);
+    Pipeline pipeline = new Pipeline()
+      .setStages(new PipelineStage[] { tokenizer, hashingTF, lr });
```
ditto

yinxusen commented on examples/src/main/python/ml/estimator_transformer_param_example.py:
```diff
@@ -0,0 +1,95 @@
+# $example on$
+from pyspark.ml.classification import LogisticRegression
+# $example off$
+from pyspark.ml.param import Param, Params
```
remote the import

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/EstimatorTransformerParamExample.scala:
```diff
@@ -0,0 +1,98 @@
+package org.apache.spark.examples.ml
+
+import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.ml.classification.LogisticRegression
```
Add necessary imports with `// $example on$` `// $example off$`

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/EstimatorTransformerParamExample.scala:
```diff
@@ -0,0 +1,98 @@
+    // We may alternatively specify parameters using a ParamMap,
+    // which supports several methods for specifying parameters.
+    val paramMap = ParamMap(lr.maxIter -> 20)
+      .put(lr.maxIter, 30) // Specify 1 Param.  This overwrites the original maxIter.
```
2-indent spaces for inline comments

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/EstimatorTransformerParamExample.scala:
```diff
@@ -0,0 +1,98 @@
+    // which supports several methods for specifying parameters.
+    val paramMap = ParamMap(lr.maxIter -> 20)
+      .put(lr.maxIter, 30) // Specify 1 Param.  This overwrites the original maxIter.
+      .put(lr.regParam -> 0.1, lr.threshold -> 0.55) // Specify multiple Params.
```
ditto

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/EstimatorTransformerParamExample.scala:
```diff
@@ -0,0 +1,98 @@
+      .put(lr.regParam -> 0.1, lr.threshold -> 0.55) // Specify multiple Params.
+
+    // One can also combine ParamMaps.
+    val paramMap2 = ParamMap(lr.probabilityCol -> "myProbability") // Change output column name
```
ditto

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/ModelSelectionViaCrossValidationExample.scala:
```diff
@@ -0,0 +1,109 @@
+package org.apache.spark.examples.ml
+
+import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.ml.Pipeline
```
add imports with `// $example on$` and `// $example off$`

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/ModelSelectionViaCrossValidationExample.scala:
```diff
@@ -0,0 +1,109 @@
+      .setEstimator(pipeline)
+      .setEvaluator(new BinaryClassificationEvaluator)
+      .setEstimatorParamMaps(paramGrid)
+      .setNumFolds(2) // Use 3+ in practice
```
2-indent spaces for inline comments

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/ModelSelectionViaTrainValidationSplitExample.scala:
```diff
@@ -0,0 +1,69 @@
+package org.apache.spark.examples.ml
+
+import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.ml.evaluation.RegressionEvaluator
```
add imports with `// $example on$` and `// $example off$`

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/PipelineExample.scala:
```diff
@@ -0,0 +1,91 @@
+package org.apache.spark.examples.ml
+
+import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.ml.{Pipeline, PipelineModel}
```
add imports with `// $example on$` and `// $example off$`

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/PipelineExample.scala:
```diff
@@ -0,0 +1,91 @@
+    // Fit the pipeline to training documents.
+    val model = pipeline.fit(training)
+
+    // now we can optionally save the fitted pipeline to disk
```
now -> Now

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/PipelineExample.scala:
```diff
@@ -0,0 +1,91 @@
+    // now we can optionally save the fitted pipeline to disk
+    model.save("/tmp/spark-logistic-regression-model")
+
+    // we can also save this unfit pipeline to disk
```
we -> We

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/PipelineExample.scala:
```diff
@@ -0,0 +1,91 @@
+    // we can also save this unfit pipeline to disk
+    pipeline.save("/tmp/unfit-lr-model")
+
+    // and load it back in during production
```
and -> And

yinxusen commented on examples/src/main/python/ml/estimator_transformer_param_example.py:
```diff
@@ -0,0 +1,95 @@
+    # parameter previously.
+    prediction = model2.transform(test)
+    selected = prediction.select(
+        "features", "label", "myProbability", "prediction")
```
no need to break the line here

yinxusen commented on examples/src/main/python/ml/pipeline_example.py:
```diff
@@ -0,0 +1,68 @@
+    # stages: tokenizer, hashingTF, and lr.
+    tokenizer = Tokenizer(inputCol="text", outputCol="words")
+    hashingTF = HashingTF(
+        inputCol=tokenizer.getOutputCol(), outputCol="features")
```
ditto

yinxusen commented on examples/src/main/python/ml/pipeline_example.py:
```diff
@@ -0,0 +1,68 @@
+        (2L, "spark f g h", 1.0),
+        (3L, "hadoop mapreduce", 0.0)], ["id", "text", "label"])
+
+    # Configure an ML pipeline, which consists of tree
```
1. three stages
2. try to fit them in a single line

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/ModelSelectionViaTrainValidationSplitExample.scala:
```diff
@@ -0,0 +1,69 @@
+    val data = sqlContext.read.format("libsvm").load("data/mllib/sample_libsvm_data.txt")
+    val Array(training, test) = data.randomSplit(Array(0.9, 0.1), seed = 12345)
+
+    val lr = new org.apache.spark.ml.regression.LinearRegression()
```
add `import org.apache.spark.ml.regression.LinearRegression` at the beginning and change here to `val lr = new LinearRegression()`

yinxusen commented:
@devaraj-kavali One more thing, could you make the two python examples identical with previous ones in markdown file? There are some comments with early indentations. Note that in Spark, the maximum characters in a line for Python is also 100, not 79 or 80 in other settings.

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaModelSelectionViaTrainValidationSplitExample.java:
```diff
@@ -0,0 +1,82 @@
+    SQLContext jsql = new SQLContext(sc);
+
+    // $example on$
+    DataFrame data = jsql.read().format("libsvm").load("data/mllib/sample_libsvm_data.txt");
```
According to the previous example, here the path should be `data/mllib/sample_linear_regression_data.txt`

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/ModelSelectionViaTrainValidationSplitExample.scala:
```diff
@@ -0,0 +1,69 @@
+
+    // $example on$
+    // Prepare training and test data.
+    val data = sqlContext.read.format("libsvm").load("data/mllib/sample_libsvm_data.txt")
```
Here the path should be `data/mllib/sample_linear_regression_data.txt`

yinxusen commented on examples/src/main/python/ml/pipeline_example.py:
```diff
@@ -0,0 +1,68 @@
+
+    # $example on$
+    # Prepare training documents from a list of (id, text, label) tuples.
+    LabeledDocument = Row("id", "text", "label")
```
I think we can remove the line. Because no line after here uses `LabeledDocument`. You can try to test whether the example runnable or not if we remove it.

yinxusen commented on examples/src/main/python/ml/pipeline_example.py:
```diff
@@ -0,0 +1,68 @@
+from pyspark.ml.classification import LogisticRegression
+# $example off$
+from pyspark.ml.feature import HashingTF, Tokenizer
+from pyspark.sql import Row
```
remove the import if you remove the `LabeledDocument` line below.

Commit:
Review comments fix about code style

SparkQA commented:
**[Test build #51472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51472/consoleFull)** for PR 11053 at commit [`ea6e77c`](https://github.com/apache/spark/commit/ea6e77cfbc5948a57e5bccfd4fbf3f57a5f9792e).

SparkQA commented:
**[Test build #51472 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51472/consoleFull)** for PR 11053 at commit [`ea6e77c`](https://github.com/apache/spark/commit/ea6e77cfbc5948a57e5bccfd4fbf3f57a5f9792e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51472/
Test PASSed.

devaraj-kavali commented:
Thanks @yinxusen for your details review and comments. I have addressed them.

yinxusen commented on examples/src/main/java/org/apache/spark/examples/ml/JavaModelSelectionViaTrainValidationSplitExample.java:
```diff
@@ -0,0 +1,82 @@
+    SQLContext jsql = new SQLContext(sc);
+
+    // $example on$
+    DataFrame data = jsql.read().format("libsvm").load("data/mllib/sample_linear_regression_data.txt");
```
Exceed 100 chars limitation

yinxusen commented on examples/src/main/python/ml/estimator_transformer_param_example.py:
```diff
@@ -0,0 +1,87 @@
+Estimator Transformer Param Example.
+"""
+from pyspark import SparkContext, SQLContext
+from pyspark.mllib.linalg import Vectors
```
The `Vectors` should be included in `$example on$`

yinxusen commented on examples/src/main/python/ml/pipeline_example.py:
```diff
@@ -0,0 +1,64 @@
+"""
+Pipeline Example.
+"""
+from pyspark import SparkContext, SQLContext
```
Change it to 

```scala
from pyspark import SparkContext, SQLContext
# $example on$
from pyspark.ml import Pipeline
from pyspark.ml.classification import LogisticRegression
from pyspark.ml.feature import HashingTF, Tokenizer
# $example off$
```

yinxusen commented on docs/ml-guide.md:
```diff
@@ -864,91 +304,12 @@ The `ParamMap` which produces the best evaluation metric is selected as the best
-  .show()
-
+{% include_example scala/org/apache/spark/examples/ml/ModelSelectionViaTrainValidationSplitExample.scala %}
 {% endhighlight %}
```
remove the line

yinxusen commented:
@devaraj-kavali All look good to me except those minors I pointed out above.

@mengxr The `ModelSelectionViaTrainValidationSplitExample` and `JavaModelSelectionViaTrainValidationSplitExample`, which call `LinearRegression` then `CholeskyDecomposition.solve()`, fail with return value 1. I am not sure why those errors occur, can you help checking it? I can run `WeightedLeastSquareSuite` successfully.

Commit:
Review comments fix

SparkQA commented:
**[Test build #51523 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51523/consoleFull)** for PR 11053 at commit [`e20c920`](https://github.com/apache/spark/commit/e20c920616347b58bb13f7ce9b027620547276a6).

SparkQA commented:
**[Test build #51523 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51523/consoleFull)** for PR 11053 at commit [`e20c920`](https://github.com/apache/spark/commit/e20c920616347b58bb13f7ce9b027620547276a6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51523/
Test PASSed.

devaraj-kavali commented:
Thanks again @yinxusen for the review, I have addressed the comments.

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/PipelineExample.scala:
```diff
@@ -0,0 +1,93 @@
+    val model = pipeline.fit(training)
+
+    // Now we can optionally save the fitted pipeline to disk
+    model.save("/tmp/spark-logistic-regression-model")
```
One more thing, it's better to make it overridable: `model.write.overwrite().save("/tmp/spark-logistic-regression-model")`, otherwise the example can only run once.

yinxusen commented on examples/src/main/scala/org/apache/spark/examples/ml/PipelineExample.scala:
```diff
@@ -0,0 +1,93 @@
+    model.save("/tmp/spark-logistic-regression-model")
+
+    // We can also save this unfit pipeline to disk
+    pipeline.save("/tmp/unfit-lr-model")
```
Ditto: `pipeline.write.overwrite().save("/tmp/unfit-lr-model")`

yinxusen commented:
@mengxr LGTM except for the minor issue above.

`ModelSelectionViaTrainValidationSplitExample` and `JavaModelSelectionViaTrainValidationSplitExample` still have a problem of Vector serialization. But I think we can add follow-up JIRA to locate the bug and fix it.

Commit:
Updated to overwrite the files to support running the example multiple
times without deleting the files

SparkQA commented:
**[Test build #51599 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51599/consoleFull)** for PR 11053 at commit [`2fe0667`](https://github.com/apache/spark/commit/2fe06672cc9827545fe34a22fc38f2e304243205).

SparkQA commented:
**[Test build #51599 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51599/consoleFull)** for PR 11053 at commit [`2fe0667`](https://github.com/apache/spark/commit/2fe06672cc9827545fe34a22fc38f2e304243205).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51599/
Test PASSed.

devaraj-kavali commented:
Thanks @yinxusen for the good suggestion, I have addressed it.

> ModelSelectionViaTrainValidationSplitExample and JavaModelSelectionViaTrainValidationSplitExample still have a problem of Vector serialization. But I think we can add follow-up JIRA to locate the bug and fix it.

Yes, we can create an another followup JIRA to fix the problem. Thank you.

mengxr commented:
Merged into master. Thanks!

@devaraj-kavali @yinxusen I guess there are some duplicate example code under `examples/ml`. Could you create a JIRA and think about possible merges? Please also create a JIRA for the serialization issue and paste the JIRA numbers here.

yinxusen commented:
@mengxr I'll do it soon.

2016年2月22日星期一，asfgit <notifications@github.com> 写道：

> Closed #11053 <https://github.com/apache/spark/pull/11053> via 02b1fef
> <https://github.com/apache/spark/commit/02b1fefffb00d50c1076a26f2f3f41f3c1fa0001>
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/11053#event-560755248>.
>


-- 
Cheers
-----------------------------------
Xusen Yin    (尹绪森)
LinkedIn: https://cn.linkedin.com/in/xusenyin

yinxusen commented:
Try to solve the serialization issue in this JIRA: https://issues.apache.org/jira/browse/SPARK-13462

yinxusen commented:
@mengxr For code cleanup and merge, let's do it after all other example code JIRAs are merged. JIRA here: https://issues.apache.org/jira/browse/SPARK-13461

devaraj-kavali commented:
@yinxusen I will look into the issue SPARK-13462, Thanks for creating it.
